### PR TITLE
[SPARK-5016][MLLib] Distribute GMM mixture components to executors

### PR DESCRIPTION
**Pull request by feynmanliang**
Distribute expensive portions of computation for Gaussian mixture components (in particular, pre-computation of `MultivariateGaussian.rootSigmaInv`, the inverse covariance matrix and covariance determinant) across executors. Repost of PR#4654.

Notes for reviewers:
- What should be the policy for when to distribute computation. Always? When numClusters > threshold? User-specified param?

TODO:
- Performance testing and comparison for large number of clusters

### CODE REVIEW

**Comment by MechCoder** [comment]
Thanks for picking up on this. Next time it might be better to just cherry-pick the commits from the branch as contributor and commit history is not lost.

**Comment by feynmanliang** [comment]
You're totally right; sorry about that! Totally slipped my mind. Will do
next time.

**Comment by feynmanliang** [comment]
I did some [perf testing](https://gist.github.com/feynmanliang/70d79c23dffc828939ec) and it shows that distributing the Gaussians does yield a significant improvement in performance when the number of clusters and dimensionality of the data is sufficiently large (>30 dimensions, >10 clusters).

In particular, the "typical" use case of 40 dimensions and 10k clusters gains about 15 seconds in runtime when distributing the Gaussians.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 1]
```diff
@@ -171,14 +185,30 @@ class GaussianMixture private (
...
+          gaussians(i) = gs(i)
+        }
+      } else {
+        var i = 0
+        while (i < k) {
+          val mu = sums.means(i) / sums.weights(i)
+          BLAS.syr(-sums.weights(i), Vectors.fromBreeze(mu),
+            Matrices.fromBreeze(sums.sigmas(i)).asInstanceOf[DenseMatrix])
+          weights(i) = sums.weights(i) / sumWeights
+          gaussians(i) = new MultivariateGaussian(mu, sums.sigmas(i) / sums.weights(i))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: This is essentially the same code as the distributed branch. Should we define a method?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 2]
```diff
@@ -121,6 +124,17 @@ class GaussianMixture private (
    */
   def getConvergenceTol: Double = convergenceTol
 
+  /**
+   * Sets indicator for computing Gaussians on driver or executors
+   */
+  def setDistributeGaussians(distributeGaussians: Boolean): this.type = {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: I'm not sure whether we should expose this flag. It should be easy to calculate the computation cost and communication cost and decide whether we want to distribute the computation. A simple heuristic using a rough estimate should work well in practice.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 3]
```diff
@@ -171,14 +185,30 @@ class GaussianMixture private (
...
-      while (i < k) {
-        val mu = sums.means(i) / sums.weights(i)
-        BLAS.syr(-sums.weights(i), Vectors.fromBreeze(mu),
-          Matrices.fromBreeze(sums.sigmas(i)).asInstanceOf[DenseMatrix])
-        weights(i) = sums.weights(i) / sumWeights
-        gaussians(i) = new MultivariateGaussian(mu, sums.sigmas(i) / sums.weights(i))
-        i = i + 1
+
+      if (distributeGaussians) {
+        val (ws, gs) = sc.parallelize(0 until k).map { i =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: `sc.parallelize` without number of partitions would use the default number of partitions, which is 2 in Scala. We can use `math.min(k, 1024)` here.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 2 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK, using the heuristic dimension >= 30 and numClusters >= 10 based on initial [perf tests](https://gist.github.com/feynmanliang/70d79c23dffc828939ec)

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 3 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 1 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK.

**Comment by feynmanliang** [comment]
jenkins retest this please

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 4]
```diff
@@ -140,6 +143,9 @@ class GaussianMixture private (
     // Get length of the input vectors
     val d = breezeData.first().length
 
+    // Logic for when to distribute the computation of the [[MultivariateGaussian]]s
+    val distributeGaussians = d >= 30 && k >= 10
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: The computation cost is `C1 * k * d^3` and the communication cost for distributing Gaussian is `C2 * k * d^2`, where `C1` and `C2` are constants. So if we run on local (assuming single thread), the total time is C1 * k * d^3, if we distribute the work, the total time is `C1 * d^3 + C2 * k * d^2`. So we should distribute the work when `(k - 1) / k * d > C2 / C1`, which is roughly controlled by `d` for a non-trivial `k`. Based on your performance test, `C2 / C1` is roughly 25. So we can use `(k - 1.0) / k * d > 25` here. The logic you suggested didn't cover the case when k is small but d is very large.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 5]
```diff
@@ -171,14 +177,25 @@ class GaussianMixture private (
...
-        val mu = sums.means(i) / sums.weights(i)
-        BLAS.syr(-sums.weights(i), Vectors.fromBreeze(mu),
-          Matrices.fromBreeze(sums.sigmas(i)).asInstanceOf[DenseMatrix])
-        weights(i) = sums.weights(i) / sumWeights
-        gaussians(i) = new MultivariateGaussian(mu, sums.sigmas(i) / sums.weights(i))
-        i = i + 1
+
+      if (distributeGaussians) {
+        val (ws, gs) = sc.parallelize(0 until k, math.min(k, 1024)).map { i =>
+          updateWeightsAndGaussians(sums.means(i), sums.sigmas(i), sums.weights(i), sumWeights)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: This might result broadcasting sums to all partitions. Could you try

~~~scala
val tuples = Seq.tabulate(i => (sums.means(i), sums.sigmas(i), sums.weights(i)))
sc.parallelize(tuples, math.min(k, 1024)).map { case (mean, sigma, weight) =>
  updateWeightsAndGaussians(mean, sigma, weight, sumWeighs)
}
~~~

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 6]
```diff
@@ -171,14 +177,25 @@ class GaussianMixture private (
...
+
+        }.collect.unzip
+        (0 until k).foreach { i =>
+          weights(i) = ws(i)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: `Array.copy(ws, 0, weights, 0, ws.length)`

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/GaussianMixtureSuite.scala [hunk 7]
```diff
@@ -106,7 +106,7 @@ class GaussianMixtureSuite extends SparkFunSuite with MLlibTestSparkContext {
       Array(0.5, 0.5),
       Array(
         new MultivariateGaussian(Vectors.dense(-1.0), Matrices.dense(1, 1, Array(1.0))),
-        new MultivariateGaussian(Vectors.dense(1.0), Matrices.dense(1, 1, Array(1.0)))
+        new MultivariateGaussian(Vectors.dense(2.0), Matrices.dense(1, 1, Array(1.0)))
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/GaussianMixtureSuite.scala]: Why do we need this change?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 4 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK, will change but will also account for number worker = math.min(k, 1024)

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 5 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: Ok

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 6 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK

File: mllib/src/test/scala/org/apache/spark/mllib/clustering/GaussianMixtureSuite.scala (hunk 7 above)
Comment (feynmanliang) [inline mllib/src/test/scala/org/apache/spark/mllib/clustering/GaussianMixtureSuite.scala]: Typo, revereted.

**Comment by feynmanliang** [comment]
jenkins retest this please

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 8]
```diff
@@ -140,6 +143,10 @@ class GaussianMixture private (
     // Get length of the input vectors
     val d = breezeData.first().length
 
+    // Logic for when to distribute the computation of the [[MultivariateGaussian]]s
+    val numPartitions = math.min(k, 1024)
+    val distributeGaussians = (k - k.toDouble / numPartitions) / numPartitions * d > 25
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: `numPartitions` might not equal to number of available executors. We can simply assume that we have enough executors to distribute the work. If `k` is large, `(k - 1) / k` is very close to `1` and hence we don't really care. So let's still keep `(k - 1) / k * d > 25` (better with comments about the heuristic), which is essentially `d > 25` but takes care of the case when `k` is really small.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 9]
```diff
@@ -43,6 +43,9 @@ import org.apache.spark.util.Utils
  * While this process is generally guaranteed to converge, it is not guaranteed
  * to find a global optimum.
  *
+ * Computation of the Gaussian pdfs are distributed across workers when dimensionality of the input
+ * data >= 30 and number of centers >= 10.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: This needs update. I don't think we need to mention the explicit rule we use in public doc. This is implementation detail.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 10]
```diff
@@ -192,6 +209,19 @@ class GaussianMixture private (
...
 
+  private def updateWeightsAndGaussians(
+      mean: BDV[Double],
+      sigma: BreezeMatrix[Double],
+      weight: Double,
+      sumWeights: Double): (Double, MultivariateGaussian) = {
+    val mu = mean / weight
+    BLAS.syr(-weight, Vectors.fromBreeze(mu),
+      Matrices.fromBreeze(sigma).asInstanceOf[DenseMatrix])
+    val newWeight = weight / sumWeights
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: Shouldn't use `sumWeights`.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 11]
```diff
@@ -171,14 +178,24 @@ class GaussianMixture private (
...
+        val (ws, gs) = sc.parallelize(tuples, numPartitions).map { case (mean, sigma, weight) =>
+          updateWeightsAndGaussians(mean, sigma, weight, sumWeights)
+        }.collect.unzip
+        Array.copy(ws, 0, weights, 0, ws.length)
+        Array.copy(gs, 0, gaussians, 0, gs.length)
+      } else {
+        var i = 0
+        while (i < k) {
+          val (weight, gaussian) =
+            updateWeightsAndGaussians(sums.means(i), sums.sigmas(i), sums.weights(i), sumWeights)
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: `sums.weights(i) / sumWeights`

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 8 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 9 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala [hunk 12]
```diff
@@ -192,6 +209,19 @@ class GaussianMixture private (
   /** Java-friendly version of [[run()]] */
   def run(data: JavaRDD[Vector]): GaussianMixtureModel = run(data.rdd)
 
+  private def updateWeightsAndGaussians(
+      mean: BDV[Double],
+      sigma: BreezeMatrix[Double],
+      weight: Double,
+      sumWeights: Double): (Double, MultivariateGaussian) = {
+    val mu = mean / weight
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: This could be in-place update: `val mu = (mean /= weight)`.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 11 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: Discussed on Slack.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 10 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: Discussed on Slack.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 12 above)
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: OK

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala (hunk 11 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/GaussianMixture.scala]: Had an offline discussion with @feynmanliang. I misunderstood the weights normalization. So we revert the change.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!
